ykllvmwrap/src/jitbuilder.cc

```diff
@@ -0,0 +1,298 @@
+        for (auto Arg = CI->arg_begin() + 1; Arg != CI->arg_end(); Arg++) {
+          Vec.push_back(Arg->get());
+        }
+        found = true;
+        break;
+      }
+    }
+  }
+  if (!found)
+    errx(EXIT_FAILURE, "failed to find trace inputs");
+  return Vec;
+}
+
+class JITBuilder {
+public:
+  ValueToValueMapTy VMap;
+#ifndef NDEBUG
+  ValueToValueMapTy RevVMap;
+#endif
+  string TraceName;
+
+  Module *createModule(char *FuncNames[], size_t BBs[], size_t Len,
+                       Module *AOTMod) {
+    LLVMContext &JITContext = AOTMod->getContext();
+    auto JITMod = new Module("", JITContext);
+    uint64_t TraceIdx = NextTraceIdx.fetch_add(1);
+    if (TraceIdx == numeric_limits<uint64_t>::max())
+      errx(EXIT_FAILURE, "trace index counter overflowed");
+
+    // Get var args from start_tracing call.
+    auto Inputs = getTraceInputs(AOTMod->getFunction(FuncNames[0]), BBs[0]);
+
+    std::vector<Type *> InputTypes;
+    for (auto Val : Inputs) {
+      InputTypes.push_back(Val->getType());
+    }
+
+    // Create function to store compiled trace.
+    TraceName = string(TRACE_FUNC_PREFIX) + to_string(TraceIdx);
+    llvm::FunctionType *FType =
+        llvm::FunctionType::get(Type::getVoidTy(JITContext), InputTypes, false);
+    llvm::Function *JITFunc = llvm::Function::Create(
+        FType, Function::InternalLinkage, TraceName, JITMod);
+    JITFunc->setCallingConv(CallingConv::C);
+
+    // Create entry block and setup builder.
+    auto DstBB = BasicBlock::Create(JITContext, "", JITFunc);
+    llvm::IRBuilder<> Builder(JITContext);
+    Builder.SetInsertPoint(DstBB);
+
+    // Variables that are used (but not defined) inbetween start and stop
+    // tracing need to be replaced with function arguments which the user passes
+    // into the compiled trace. This loop creates a mapping from those original
+    // variables to the function arguments of the compiled trace function.
+    for (size_t Idx = 0; Idx != Inputs.size(); Idx++) {
+      Value *OldVal = Inputs[Idx];
+      Value *NewVal = JITFunc->getArg(Idx);
+      assert(NewVal->getType()->isPointerTy());
+      VMap[OldVal] = NewVal;
+    }
+
+    // A pointer to the call to YKTRACE_START in the AOT module (once
+    // encountered). When this changes from NULL to non-NULL, then we start
+    // copying instructions from the AOT module into the JIT module.
+    Instruction *StartTracingInstr = nullptr;
+
+    std::vector<CallInst *> inlined_calls;
+    CallInst *last_call = nullptr;
+    std::vector<GlobalVariable *> cloned_globals;
+
+    // Iterate over the PT trace and stitch together all traced blocks.
+    for (size_t Idx = 0; Idx < Len; Idx++) {
+      auto FuncName = FuncNames[Idx];
+
+      // FIXME: Deal with holes in the trace, e.g. calls to libc.
+      assert(FuncName != nullptr);
+
+      // Get a traced function so we can extract blocks from it.
+      Function *F = AOTMod->getFunction(FuncName);
+      if (!F)
```

Comment:
Would it be easier to use `assert`s for things like these that, I assume, we don't think can ever go wrong?

Reply:
It depends if you want a good error message or not in release mode (where C++ assertions are not compiled in). If `getFunction()` returned null and we'd used an assertion to check for this, in release mode we'd probably get a null pointer dereference somewhere down the line.

Personally, I think an explicit check is good here.

Reply:
This sounds like a debug assert sort of thing to me. I don't think we want to pay this penalty all the time?

[I do agree that we need a non-debug assert thing too. I don't know if one's available or whether we have to make one ourselves.]

Reply:
The problem is, in C++ there's no such thing as a debug assert. There's one kind of assertion and they are either on (debug builds) or off (release builds with `-DNDEBUG`).

We could roll our own debug assertion code, but I'd say that's out of scope of Lukas' PR.

Reply:
Agreed it's out of scope here, but why don't we fix this in a PR in the next day or two? [Before we end up with loads of not-very-useful `errx` calls, slowing down release mode, and probably not making the code clearer either.]

Reply:
I realised that I got my logic all wrong. Let me try again.

C assertions are enabled only in Cargo debug builds. Therefore an `assert()` in C++ code is in fact like a `debug_assert!`. What we don't have in our setup is *release assertions* for C++ code.

Whether or not we try to implement release assertions, this particular line of code can be an `assert()`.

Sorry for the screw-up!

Reply:
I meant the same as you, so we are now on the same page (though I assumed we were on the same page before ;)).